### PR TITLE
Adjust transaction handling via db.Context

### LINKED ISSUE #19513: Make sure if a DB Session got created for a request, it stays unique per request
to prevent deadlocks, etc...:
- [x] use context for all function calling package model and model itself
  - #19511
  - ...
- [ ] add func that can extract an open session from context if there is one or open a new one
- [ ] be able to make sure, if `db.TxContext()` was already opened, a function within the stack does not open a second one again

### PR DESCRIPTION
**Pull request by lunny**
Partially fix #19513

### CODE REVIEW

File: models/db/context.go [hunk 1]
```diff
@@ -23,39 +23,46 @@ type contextKey struct {
...
+	if key == enginedContextKey {
 		return ctx
 	}
 	return ctx.Context.Value(key)
 }
 
 // WithContext returns this engine tied to this context
 func (ctx *Context) WithContext(other context.Context) *Context {
-	return WithEngine(other, ctx.e)
+	return newContext(ctx, ctx.e.Context(ctx), ctx.transaction)
```
Comment (delvh) [inline models/db/context.go]: This doesn't seem right: You're not even using the `other` parameter...

File: models/db/context.go [hunk 2]
```diff
@@ -23,39 +23,46 @@ type contextKey struct {
...
-		Context: ctx,
-		e:       e.Context(ctx),
+		Context:     ctx,
+		e:           e,
+		transaction: transaction,
 	}
 }
 
+// InTransaction if context is in a transaction
+func (ctx *Context) InTransaction() bool {
```
Comment (delvh) [inline models/db/context.go:46]: What exactly is the purpose of this method?
It seems currently unused.
And also, when would that be good to know?

**Comment by 6543** [comment]
please resolve conflicts

File: models/db/context.go (hunk 2 above)
Comment (lunny) [reply models/db/context.go:46]: It's a requirement from #19513 , I think there are some requirements in some place but previously there is no suitable function to use.

**Review by delvh — APPROVED** [review]
So, to recap:
You opened a PR titled "Adjust something", and when asked where this something is needed, the answer boils down to "somewhere, perhaps".

No further questions. LGTM.

**Review by 6543 — APPROVED** [review]

**Pull request merged**